lib/sprintf.js

```diff
@@ -0,0 +1,187 @@
+    this.size -= 1;
+  }
+
+  var val = L.append(this.list, {key: key});
+  this.dict[key] = val;
+  this.size += 1;
+
+  return val;
+};
+
+function format(fmt /*, ... */) {
+  return vsprintf(fmt, arguments, compat);
+}
+
+function sprintf(fmt /*, ... */) {
+  return vsprintf(fmt, arguments, extended);
+}
+
+function vsprintf(fmt, argv, compile) {
+  if (typeof(fmt) !== 'string') {
+    if (compile === compat)
+      return concat(argv);
+    fmt = '' + fmt;
+  }
+
+  var lru = compile.lru;
+  var cached = lru.find(fmt);
+
+  if (cached === undefined) {
+    cached = lru.insert(fmt);
+    cached.fun = compile(fmt);
+  }
+
+  return cached.fun.apply(null, argv);
+}
+
+function concat(argv) {
+  var argc = argv.length;
+  var objects = new Array(argc);
+
+  for (var i = 0; i < argc; i += 1)
+    objects[i] = inspect(argv[i]);
+
+  return objects.join(' ');
+}
+
+// util.format() compatibility mode.
+function compat(fmt) {
+  var index = 0;
+  var argc = 0;
+  var body = '';
+
+  for (;;) {
+    var start = index;
+    index = fmt.indexOf('%', index);
+    var end = index === -1 ? fmt.length : index;
+
+    if (start !== end) {
+      body += ' s += "' + untaint(fmt.slice(start, end)) + '";';
+    }
+
+    if (index === -1)
+      break;
+
+    index += 1;
+
+    var c = fmt[index];
+    switch (c) {
+      case 'd':
+        // %d is really %f; it coerces the argument to a number
+        // before turning it into a string.
+        body += ' s += argc > ' + (argc + 1) + ' ? +a' + argc + ' : "%d";';
+        argc += 1;
+        index += 1;
+        continue;
+
+      case 'j':
+        body += ' if (argc > ' + (argc + 1) + ')';
+        body += ' try { s += JSON.stringify(a' + argc + '); }';
+        body += ' catch (e) { s += "[Circular]"; }';
+        body += ' else s += "%j";';
+        argc += 1;
+        index += 1;
+        continue;
+
+      case 's':
+        body += ' s += argc > ' + (argc + 1) + ' ? a' + argc + ' : "%s";';
+        argc += 1;
+        index += 1;
+        continue;
+
+      case '"':
+        c = '\\"';
+        // Fall through.
+
+      default:
+        body += ' s += "' + c + '";';
+        index += 1;
+        continue;
+    }
+  }
+
+  // The dummy argument lets format() and sprintf() call the formatter
+  // with .apply() without having to unshift the format argument first.
+  var source = 'return function(dummy';
+
+  for (var i = 0; i < argc; i += 1)
+    source += ', a' + i;
+
+  source += ') { var argc = arguments.length, s = "";' + body;
+  source += ' for (var i = ' + argc + ' + 1; i < arguments.length; i += 1) {';
+  source += ' var arg = arguments[i];';
+  source += ' if (arg === null || typeof(arg) !== "object") s += " " + arg;';
+  source += ' else s += " " + inspect(arg);';
+  source += ' }';
+  source += ' return s;';
+  source += ' };';
+
+  return Function('inspect', source)(inspect);
+}
+
+function untaint(s) {
+  return s.replace(/[\\"]|[^\x20-\x7F]/g, function(c) {
```

Comment:
Now that you bring it up: probably not.  With Harmony template strings being available now, the generated code is susceptible to template literal injection.  I'll have to look at this again.